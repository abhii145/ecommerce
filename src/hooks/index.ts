import axios from "axios"
import {
  keepPreviousData,
  useQuery,
  UseQueryResult,
} from "@tanstack/react-query"
import { Category, Product, UseProductsParams, ProductResponse } from "../types"

const API_BASE_URL = "https://dummyjson.com/products"

export const useCategories = (): UseQueryResult<Category[]> => {
  return useQuery({
    queryKey: ["categories"],
    queryFn: async () => {
      const response = await axios.get(`${API_BASE_URL}/categories`)
      return response.data
    },
    staleTime: 1000 * 60 * 5,
    placeholderData: keepPreviousData,
  })
}

export const useProducts = ({
  limit,
  skip,
  searchQuery,
  category,
  sortBy,
  order,
}: UseProductsParams): UseQueryResult<ProductResponse> => {
  return useQuery({
    queryKey: ["products", limit, skip, searchQuery, category, sortBy, order],
    queryFn: async () => {
      let url = `${API_BASE_URL}?limit=${limit}&skip=${skip}&sortBy=${sortBy}&order=${order}`

      if (searchQuery) {
        url = `${API_BASE_URL}/search?q=${searchQuery}&limit=${limit}&skip=${skip}`
      } else if (category && category !== "All") {
        url = `${API_BASE_URL}/category/${category}?limit=${limit}&skip=${skip}&sortBy=${sortBy}&order=${order}`
      }

      const response = await axios.get(url)
      return response.data
    },
    placeholderData: keepPreviousData,
    staleTime: 1000 * 60 * 5,
  })
}

export const useSingleProduct = (id: string): UseQueryResult<Product> => {
  return useQuery({
    queryKey: ["singleProducts", id],
    queryFn: async () => {
      const response = await axios.get(`${API_BASE_URL}/${id}`)
      return response.data
    },
    staleTime: 1000 * 60 * 5,
  })
}

export const useSearchSuggestions = (query: string): UseQueryResult<Product[]> => {
  return useQuery({
    queryKey: ["searchSuggestions", query],
    queryFn: async () => {
      const response = await axios.get(`${API_BASE_URL}/search?q=${query}`)
      return response.data.products.map((product: Product) => ({
        ...product,
        category: product.category || "Uncategorized",
      }))
    },
    enabled: !!query,
    staleTime: 1000 * 60 * 5,
  })
}
