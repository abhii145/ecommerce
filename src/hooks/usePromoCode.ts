import { useDispatch, useSelector } from "react-redux"
import { RootState } from "../store/store"
import {
  setPromoCode,
  clearPromoCode,
  applyPromoCode,
  selectPromoCode,
  clearPromoMessage,
} from "../store/promoCodeSlice"

export const usePromoCode = () => {
  const dispatch = useDispatch()
  const { promoCode, discount, promoMessage } = useSelector(selectPromoCode)
  const { orders } = useSelector((state: RootState) => state.orders)
  const cartItems = useSelector((state: RootState) => state.cart.items)

  const handleApplyPromoCode = () => {
    dispatch(applyPromoCode({ cartItems, orders }))
  }

  const handlePromoCodeChange = (
    e: React.ChangeEvent<HTMLInputElement>,
    resetSelectedCoupon: () => void
  ) => {
    dispatch(setPromoCode(e?.target?.value))
    if (e.target.value === "") {
      dispatch(clearPromoCode())
      resetSelectedCoupon()
    }
  }

  const handleCouponSelect = (couponCode: string) => {
    dispatch(setPromoCode(couponCode))
    dispatch(clearPromoMessage())
  }

  return {
    promoCode,
    discount,
    promoMessage,
    handleApplyPromoCode,
    handlePromoCodeChange,
    handleCouponSelect,
  }
}
