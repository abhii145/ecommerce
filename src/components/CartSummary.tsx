import { useState, useMemo } from "react"
import { Link } from "react-router"
import { CartSummaryProps } from "../types"
import americanExpress from "../assets/american-express.svg"
import visa from "../assets/visa.svg"
import masterCard from "../assets/mastercard.svg"
import gpay from "../assets/gpay.svg"
import { usePromoCode } from "../hooks/usePromoCode"
import { availableCoupons } from "../constants"

const CartSummary = ({ total, tax, onProceedToAddress }: CartSummaryProps) => {
  const {
    promoCode,
    discount,
    promoMessage,
    handleApplyPromoCode,
    handlePromoCodeChange,
    handleCouponSelect,
  } = usePromoCode()

  const [isCouponDrawerOpen, setCouponDrawerOpen] = useState(false)
  const [selectedCoupon, setSelectedCoupon] = useState("")

  const resetSelectedCoupon = () => setSelectedCoupon("")

  const shippingCharge = useMemo(
    () => (Number(total) + Number(tax) < 499 ? 50 : 0),
    [total, tax]
  )
  const taxAmount = useMemo(() => (Number(total) * 0.18).toFixed(2), [total])
  const finalTotal = useMemo(
    () =>
      (
        Number(total) +
        shippingCharge +
        Number(taxAmount) -
        Number(discount)
      ).toFixed(2),
    [total, shippingCharge, taxAmount, discount]
  )

  return (
    <div className="bg-white rounded-md px-4 py-6 h-max">
      <ul className="text-gray-800 space-y-4">
        {[
          { label: "Subtotal", value: total },
          { label: "Tax (18%)", value: taxAmount },
          { label: "Shipping", value: shippingCharge.toFixed(2) },
          { label: "Discount", value: `${discount}` },
        ].map((item) => (
          <li key={item.label} className="flex flex-wrap gap-4 text-sm">
            <span>{item.label}</span>
            <span
              className="ml-auto font-bold"
              aria-label={`${item.label} amount`}
            >
              ₹{item.value}
            </span>
          </li>
        ))}
        {shippingCharge > 0 && (
          <li className="text-red-500 text-sm">
            Enjoy free delivery on orders over ₹499!
          </li>
        )}
        <hr className="border-gray-300" />
        <li className="flex flex-wrap gap-4 text-sm font-bold">
          <span>Total</span>
          <span className="ml-auto" aria-label="Total amount">
            ₹{finalTotal}
          </span>
        </li>
      </ul>
      <div className="mt-4 flex items-center space-x-2">
        <label htmlFor="promoCode" className="sr-only">
          Promo Code
        </label>
        <input
          id="promoCode"
          type="text"
          placeholder="Promo Code"
          value={promoCode}
          onChange={(e) => handlePromoCodeChange(e, resetSelectedCoupon)}
          className="text-sm px-4 py-2 w-full border border-gray-300 rounded-md appearance-none focus:outline-none"
        />
        <button
          type="button"
          onClick={handleApplyPromoCode}
          className="text-sm px-2 py-2 font-semibold tracking-wide bg-green-400 hover:bg-green-500 text-white rounded-md"
        >
          Apply
        </button>
      </div>

      {promoMessage && (
        <p
          className={`mt-2 text-sm ${
            discount ? "text-green-500" : "text-red-500"
          }`}
        >
          {promoMessage}
        </p>
      )}

      <div className="mt-4">
        <button
          type="button"
          onClick={() => setCouponDrawerOpen(!isCouponDrawerOpen)}
          className="text-sm text-blue-600 hover:text-blue-800"
        >
          {isCouponDrawerOpen ? "Close Coupons" : "Best Coupons for You"}
        </button>
      </div>

      {isCouponDrawerOpen && (
        <div className="mt-4 p-4 border border-gray-300 rounded-md">
          <h3 className="text-lg font-bold">Available Coupons</h3>
          <ul className="space-y-2">
            {availableCoupons.map((coupon) => (
              <li key={coupon.code} className="flex items-center gap-2">
                <input
                  type="radio"
                  id={coupon.code}
                  name="coupon"
                  value={coupon.code}
                  checked={selectedCoupon === coupon.code}
                  onChange={() => {
                    setSelectedCoupon(coupon.code)
                    handleCouponSelect(coupon.code)
                  }}
                  className="text-sm"
                />
                <label htmlFor={coupon.code} className="text-sm">
                  <strong>{coupon.description}</strong> <br />
                  <span className="text-xs text-gray-500">{coupon.terms}</span>
                </label>
              </li>
            ))}
          </ul>
        </div>
      )}

      <div className="mt-8 space-y-2">
        <button
          type="button"
          className="text-sm px-4 py-2.5 w-full font-semibold tracking-wide bg-gray-800 hover:bg-gray-900 text-white rounded-md"
          onClick={() => onProceedToAddress(discount, promoCode)}
        >
          Proceed to Address
        </button>
        <button
          type="button"
          className="text-sm px-4 py-2.5 w-full font-semibold tracking-wide bg-transparent hover:bg-gray-100 text-gray-800 border border-gray-300 rounded-md"
        >
          <Link to="/products">Continue Shopping</Link>
        </button>
      </div>

      {/* Payment Methods */}
      <div className="mt-4 flex flex-wrap justify-center gap-4">
        <img
          src={americanExpress}
          alt="American Express"
          className="w-10 object-contain"
          loading="lazy"
        />
        <img
          src={visa}
          alt="Visa"
          className="w-10 object-contain"
          loading="lazy"
        />
        <img
          src={masterCard}
          alt="MasterCard"
          className="w-10 object-contain"
          loading="lazy"
        />
        <img
          src={gpay}
          alt="Google Pay"
          className="w-10 object-contain"
          loading="lazy"
        />
      </div>
    </div>
  )
}

export default CartSummary
