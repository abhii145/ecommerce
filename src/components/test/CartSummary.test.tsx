import { render, screen, fireEvent } from "@testing-library/react"
import CartSummary from "../CartSummary"
import { BrowserRouter as Router } from "react-router"
import { usePromoCode } from "../../hooks/usePromoCode"

jest.mock("../../hooks/usePromoCode")

const mockUsePromoCode = usePromoCode as jest.MockedFunction<
  typeof usePromoCode
>

describe("CartSummary", () => {
  beforeEach(() => {
    mockUsePromoCode.mockReturnValue({
      promoCode: "",
      discount: 0,
      promoMessage: "",
      handleApplyPromoCode: jest.fn(),
      handlePromoCodeChange: jest.fn(),
      handleCouponSelect: jest.fn(),
    })
  })

  const setup = (props = {}) => {
    const defaultProps = {
      total: "1000",
      tax: "180",
      onProceedToAddress: jest.fn(),
      ...props,
    }
    return render(
      <Router>
        <CartSummary {...defaultProps} />
      </Router>
    )
  }

  it("matches snapshot", () => {
    const { asFragment } = setup()
    expect(asFragment()).toMatchSnapshot()
  })

  it("renders CartSummary component", () => {
    setup()
    expect(screen.getByText("Subtotal")).toBeInTheDocument()
    expect(screen.getByText("Tax (18%)")).toBeInTheDocument()
    expect(screen.getByText("Shipping")).toBeInTheDocument()
    expect(screen.getByText("Discount")).toBeInTheDocument()
    expect(screen.getByText("Total")).toBeInTheDocument()
  })

  it("calculates shipping charge correctly", () => {
    setup({ total: "400", tax: "72" })
    expect(screen.getByText("₹50.00")).toBeInTheDocument()
    expect(
      screen.getByText("Enjoy free delivery on orders over ₹499!")
    ).toBeInTheDocument()
  })

  it("renders shipping message correctly based on total", () => {
    setup({ total: "400", tax: "72" })
    expect(screen.getByText("₹50.00")).toBeInTheDocument()
    expect(
      screen.getByText("Enjoy free delivery on orders over ₹499!")
    ).toBeInTheDocument()
  })

  it("does not render free shipping message when total is over 499", () => {
    setup({ total: "500", tax: "90" })
    expect(screen.queryByText("₹50.00")).not.toBeInTheDocument()
    expect(
      screen.queryByText("Enjoy free delivery on orders over ₹499!")
    ).not.toBeInTheDocument()
  })

  it("applies promo code", () => {
    const handleApplyPromoCode = jest.fn()
    mockUsePromoCode.mockReturnValue({
      promoCode: "SAVE10",
      discount: 100,
      promoMessage: "Promo code applied successfully!",
      handleApplyPromoCode,
      handlePromoCodeChange: jest.fn(),
      handleCouponSelect: jest.fn(),
    })
    setup()
    fireEvent.click(screen.getByText("Apply"))
    expect(handleApplyPromoCode).toHaveBeenCalled()
    expect(
      screen.getByText("Promo code applied successfully!")
    ).toBeInTheDocument()
  })

  it("toggles coupon drawer", () => {
    setup()
    const button = screen.getByText("Best Coupons for You")
    fireEvent.click(button)
    expect(screen.getByText("Close Coupons")).toBeInTheDocument()
    fireEvent.click(screen.getByText("Close Coupons"))
    expect(screen.getByText("Best Coupons for You")).toBeInTheDocument()
  })

  it("proceeds to address", () => {
    const onProceedToAddress = jest.fn()
    setup({ onProceedToAddress })
    fireEvent.click(screen.getByText("Proceed to Address"))
    expect(onProceedToAddress).toHaveBeenCalledWith(0, "")
  })

  it("renders payment methods", () => {
    setup()
    expect(screen.getByAltText("American Express")).toBeInTheDocument()
    expect(screen.getByAltText("Visa")).toBeInTheDocument()
    expect(screen.getByAltText("MasterCard")).toBeInTheDocument()
    expect(screen.getByAltText("Google Pay")).toBeInTheDocument()
  })

  it("handles promo code change", () => {
    const handlePromoCodeChange = jest.fn()
    mockUsePromoCode.mockReturnValue({
      promoCode: "",
      discount: 0,
      promoMessage: "",
      handleApplyPromoCode: jest.fn(),
      handlePromoCodeChange,
      handleCouponSelect: jest.fn(),
    })
    setup()
    const input = screen.getByPlaceholderText("Promo Code")
    fireEvent.change(input, { target: { value: "NEWCODE" } })
    expect(handlePromoCodeChange).toHaveBeenCalled()
  })

  it("resets selected coupon on promo code change", () => {
    const handlePromoCodeChange = jest.fn()
    mockUsePromoCode.mockReturnValue({
      promoCode: "",
      discount: 0,
      promoMessage: "",
      handleApplyPromoCode: jest.fn(),
      handlePromoCodeChange,
      handleCouponSelect: jest.fn(),
    })
    setup()
    const input = screen.getByPlaceholderText("Promo Code")
    fireEvent.change(input, { target: { value: "NEWCODE" } })
    expect(handlePromoCodeChange).toHaveBeenCalled()
  })
})
