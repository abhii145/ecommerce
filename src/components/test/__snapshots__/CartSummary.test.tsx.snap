// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartSummary matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="bg-white rounded-md px-4 py-6 h-max"
  >
    <ul
      class="text-gray-800 space-y-4"
    >
      <li
        class="flex flex-wrap gap-4 text-sm"
      >
        <span>
          Subtotal
        </span>
        <span
          aria-label="Subtotal amount"
          class="ml-auto font-bold"
        >
          ₹1000
        </span>
      </li>
      <li
        class="flex flex-wrap gap-4 text-sm"
      >
        <span>
          Tax (18%)
        </span>
        <span
          aria-label="Tax (18%) amount"
          class="ml-auto font-bold"
        >
          ₹180.00
        </span>
      </li>
      <li
        class="flex flex-wrap gap-4 text-sm"
      >
        <span>
          Shipping
        </span>
        <span
          aria-label="Shipping amount"
          class="ml-auto font-bold"
        >
          ₹0.00
        </span>
      </li>
      <li
        class="flex flex-wrap gap-4 text-sm"
      >
        <span>
          Discount
        </span>
        <span
          aria-label="Discount amount"
          class="ml-auto font-bold"
        >
          ₹0
        </span>
      </li>
      <hr
        class="border-gray-300"
      />
      <li
        class="flex flex-wrap gap-4 text-sm font-bold"
      >
        <span>
          Total
        </span>
        <span
          aria-label="Total amount"
          class="ml-auto"
        >
          ₹1180.00
        </span>
      </li>
    </ul>
    <div
      class="mt-4 flex items-center space-x-2"
    >
      <label
        class="sr-only"
        for="promoCode"
      >
        Promo Code
      </label>
      <input
        class="text-sm px-4 py-2 w-full border border-gray-300 rounded-md appearance-none focus:outline-none"
        id="promoCode"
        placeholder="Promo Code"
        type="text"
        value=""
      />
      <button
        class="text-sm px-2 py-2 font-semibold tracking-wide bg-green-400 hover:bg-green-500 text-white rounded-md"
        type="button"
      >
        Apply
      </button>
    </div>
    <div
      class="mt-4"
    >
      <button
        class="text-sm text-blue-600 hover:text-blue-800"
        type="button"
      >
        Best Coupons for You
      </button>
    </div>
    <div
      class="mt-8 space-y-2"
    >
      <button
        class="text-sm px-4 py-2.5 w-full font-semibold tracking-wide bg-gray-800 hover:bg-gray-900 text-white rounded-md"
        type="button"
      >
        Proceed to Address
      </button>
      <button
        class="text-sm px-4 py-2.5 w-full font-semibold tracking-wide bg-transparent hover:bg-gray-100 text-gray-800 border border-gray-300 rounded-md"
        type="button"
      >
        <a
          data-discover="true"
          href="/products"
        >
          Continue Shopping
        </a>
      </button>
    </div>
    <div
      class="mt-4 flex flex-wrap justify-center gap-4"
    >
      <img
        alt="American Express"
        class="w-10 object-contain"
        loading="lazy"
        src="test-file-stub"
      />
      <img
        alt="Visa"
        class="w-10 object-contain"
        loading="lazy"
        src="test-file-stub"
      />
      <img
        alt="MasterCard"
        class="w-10 object-contain"
        loading="lazy"
        src="test-file-stub"
      />
      <img
        alt="Google Pay"
        class="w-10 object-contain"
        loading="lazy"
        src="test-file-stub"
      />
    </div>
  </div>
</DocumentFragment>
`;
